### ISSUE TITLE
Investigate Utopia to auto-generate OpenAPI documentation

### ISSUE THREAD

**Issue by agourlay** [qdrant/qdrant#1645]
**Is your feature request related to a problem? Please describe.**
Our current strategy to manage the OpenAPI specification is error prone and manual.

Adding new REST endpoints require to:

1. code endpoints and model in Rust
2. change specs in `/openapi/*ytt.yaml`
3. add new schema definitions to `src/schema_generator.rs`
4. run `/tools/generate_openapi_models.sh` to generate specs

Derive the documentation from the API definition would be lovely.

**Describe the solution you'd like**
Investigate auto-generating the OpenAPI documentation using the [utopia](https://github.com/juhaku/utoipa) project.

It supports natively [actix-web](https://github.com/juhaku/utoipa/blob/master/examples/actix-web-multiple-api-docs-with-scopes/src/main.rs)

**Describe alternatives you've considered**
Keeping the status quo.

**Comment by coszio** [comment]
After analyzing the `utoipa` docs, it seems straight-forward to use. The amount of action would be kinda similar, but the big difference would be that we'd use a code-first approach, so that we won't have to jump so much between files once we want to add changes to the API definition. I believe it does make it a little more fool-proof.

### What would change?
If we used this, the process would be the following:
| Current | Utoipa |
|--------|--------|
| Code endpoint and model in Rust | = |
| change specs in `/openapi/*ytt.yaml` | change the specs using proc macros |
| add new schema definitions to `src/schema_generator.rs` | add derive macro on the schemas, tie them to Api struct |
| run /tools/generate_openapi_models.sh to generate specs | specs are generated on build | 

### What's the catch?
The amount of work required to install this would consist of:
1. Add dependency (ultra quick)
2. Copy docs from current ytt files to utoipa proc macros (laborious)
3. Add derive macros on the schemas (fairly quick)
4. Setup API generation by tying the paths and schemas, and outputting to file (similar to schema_generator.rs)

**Comment by agourlay** [comment]
Thanks for the analysis :+1: 

One thing I forgot to mention is that the success criteria here is to still be able to generate our Python [client](https://github.com/qdrant/qdrant-client).

This client is the main consumer of the OpenAPI specification.

Such migration would have to be done with extra care regarding schema compatibility in order to not change too much the client.

**Comment by coszio** [comment]
Good point, the idea would be to generate an exact replica of what we have now, but using this library

**Comment by timvisee** [comment]
Though very minor, I also wonder if it will pick up things like the validation rules:

https://github.com/qdrant/qdrant/blob/4b03cf66ab365b82b3a76212d7fb10809b43317d/lib/collection/src/optimizers_builder.rs#L60

to

https://github.com/qdrant/qdrant/blob/4b03cf66ab365b82b3a76212d7fb10809b43317d/docs/redoc/master/openapi.json#L4057

**Comment by coszio** [comment]
@timvisee, great catch.
Just checked, it does not. However, it does come with its own #[schema] annotation like:
```rust
#[derive(utoipa::ToSchema)]
struct Item {
    #[schema(maximum = 10, minimum = 5, multiple_of = 2.5)]
    id: i32,
    #[schema(max_length = 10, min_length = 5, pattern = "[a-z]*")]
    value: String,
    #[schema(max_items = 5, min_items = 1)]
    items: Vec<String>,
}
```
[source](https://docs.rs/utoipa/latest/utoipa/derive.ToSchema.html#examples)

Not sure if we would use both or just one of them
